torch_xla/distributed/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,157 @@
+from copy import deepcopy
+from typing import (
+    Any,
+    Iterator,
+    Optional,
+    Type,
+)
+
+import torch
+import torch.nn as nn
+from torch import Tensor
+from torch.optim import Optimizer
+
+import torch_xla
+import torch_xla.core.xla_model as xm
+
+
+class ZeroRedundancyOptimizer(Optimizer):
+  r"""
+    ZeRO-1 wrapper. This class can wraps an arbitrary :class:`optim.Optimizer
+    <torch.optim.Optimizer>` and shards its states across ranks.
+
+    Arguments:
+        params (``Iterable``): an ``Iterable`` of :class:`torch.Tensor` s
+            or :class:`dict` s giving all parameters, which will be sharded
+            across ranks.
+        optimizer_class (:class:`torch.nn.Optimizer`): the class of the local
+            optimizer.
+        optimizer_dtype (:class:`torch.dtype`, optional): the desired data type
+            of optimizer. Default: ``torch.float32``
+        grad_clipping (bool, optional): enable (True) or disable (False) grad
+            clipping. Default: True
+        max_norm (float, optional): max norm of the gradients, effective only
+            when ``grad_clipping`` is True. Default: 1.0
+        pin_layout (bool, Optional): if ``True``, then pin the layout in the
+            collective ops (all_gather and reduce_scatter). See `xm.all_reduce`
+            for details on pinning layout. Default: True
+        **defaults: any trailing arguments, which are forwarded to the local
+            optimizer.
+    """
+
+  def __init__(
+      self,
+      params: Iterator[Tensor],
+      optimizer_class: Type[Optimizer],
+      optimizer_dtype: Optional[Any] = None,
+      grad_clipping: bool = True,
+      max_norm: Optional[float] = None,
+      pin_layout: bool = True,
+      **defaults: Any,
+  ):
+    self.params = list(params)
+    super().__init__(self.params, defaults)
+
+    self.device = self.params[0].device
+
+    self.rank = xm.get_ordinal()
+    self.world_size = xm.xrt_world_size()
+    self.cc_op_groups = [list(range(self.world_size))]
+
+    self.optimizer_dtype = optimizer_dtype if optimizer_dtype is not None else torch.float32
+    self.grad_clipping = grad_clipping
+    self.max_norm = max_norm if max_norm is not None else 1.0
+    self.pin_layout = pin_layout
+
+    # Shard parameters for use in optimizer
+    self.sharded_params = []
+    self._shard_parameters()
+    # Optimizer initialization
+    self.base_optimizer = optimizer_class(iter(self.sharded_params), **defaults)
+
+  def _shard_tensor(self, tensor: torch.Tensor):
+    """
+        Get the shard of the input tensor.
+        """
+    assert tensor.shape[0] % self.world_size == 0, "Not support padding now."
+    tensor = tensor.chunk(self.world_size)[self.rank]
+    return tensor
+
+  def _shard_parameters(self):
+    """
+        Shard all parameters.
+        """
+    xm.unlazy(self.params)
+    for param in self.params:
+      shard_data = param.data.to(device="cpu")  # move to cpu
```

Comment:
It's a mistake that it has no impact on num of graphs (1 vs 1).
In our use case, usually there is a `mark_step` after model init, this will generate one graph:
```python
model, optimizer = init_model_and_optimizer(...)
xm.mark_step()
```
as shard depends on rank, if we shard on xla device, each rank will generate a different graph. by moving this to cpu, we can share only one compiled graph between processes